strawberryfields/decompositions.py

```diff
@@ -632,6 +633,276 @@ def triangular(V, tol=1e-11):
+    zeta = -np.angle(V[0, 0])
+    V = V @ P(0, zeta, m)
+    phases["zetas"][0] = zeta
+
+    for j in range(1, m):
+        zeta = np.angle(V[0, 0]) - np.angle(V[j, j])
+        V = V @ P(j, zeta, m)
+        phases["zetas"][j] = zeta
+
+    assert np.allclose(V, np.eye(m), rtol=rtol, atol=atol), "decomposition failed"
+
+    return phases
+
+
+def _absorb_zeta(phases):
+    r"""Adjust rectangular decomposition to relocate residual phase-shifters of interferometer to edge-shifters, as given in FIG. 4 and "Relocating residual phase-shifts" section of (arXiv:2104.0756).
+
+    Args:
+        phases (dict): output of _rectangular_compact_init
+
+    Returns:
+        phases (dict): A dictionary containing the following items:
+
+        * ``m``: the length of the matrix
+        * ``phi_ins``: parameter of the phase-shifter at the beginning of the mode
+        * ``sigmas``: parameter of the sMZI :math:`\frac{(\theta_1+\theta_2)}{2}`, where :math:`\theta_{1,2}` are the values of the two internal phase-shifts of sMZI
+        * ``deltas``: parameter of the sMZI :math:`\frac{(\theta_1-\theta_2)}{2}`, where :math:`\theta_{1,2}` are the values of the two internal phase-shifts of sMZI
+                * ``phi_edges``: parameters of the edge phase shifters
+        * ``phi_outs``: parameter of the phase-shifter at the end of the mode
+
+    """
+    m = phases["m"]
+    new_phases = phases.copy()
+    del new_phases["zetas"]
+    new_phases["phi_edges"] = defaultdict(float)  # (mode, layer) : phi
+
+    if m % 2 == 0:
+        new_phases["phi_outs"][0] = phases["zetas"][0]
+        for j in range(1, m):
+            zeta = phases["zetas"][j]
+            layer = m - j
+            for mode in range(j, m - 1, 2):
+                new_phases["sigmas"][mode, layer] += zeta
+            for mode in range(j + 1, m - 1, 2):
+                new_phases["sigmas"][mode, layer - 1] -= zeta
+            if layer % 2 == 1:
+                new_phases["phi_edges"][m - 1, layer] += zeta
+            else:
+                new_phases["phi_edges"][m - 1, layer - 1] -= zeta
+    else:
+        for j in range(m):
+            zeta = phases["zetas"][j]
+            layer = m - j - 1
+            for mode in range(j, m - 1, 2):
+                new_phases["sigmas"][mode, layer] += zeta
+            for mode in range(j + 1, m - 1, 2):
+                new_phases["sigmas"][mode, layer - 1] -= zeta
+            if layer % 2 == 0:
+                new_phases["phi_edges"][m - 1, layer] += zeta
+            else:
+                new_phases["phi_edges"][m - 1, layer - 1] -= zeta
+    return new_phases
+
+
+def rectangular_compact(U, rtol=1e-12, atol=1e-12):
+    r"""Rectangular decomposition of a unitary with sMZIs and phase-shifters, as given in FIG. 3+4 and "The Clements Scheme" section of (arXiv:2104.0756).
+
+    Args:
+        U (array): unitary matrix
+
+    Returns:
+        phasess (dict): A dictionary containing the following items:
```

Comment:
```suggestion
        dict: A dictionary containing the following items:
```
With the `Returns:` section, only a type is given, since Python functions don't name the return statement :)